networkx/algorithms/shortest_paths/weighted.py

```diff
@@ -1320,7 +1338,11 @@ def _bellman_ford(
                     # therefore u is always in the dict recent_update
                     if heuristic:
                         if v in recent_update[u]:
-                            raise nx.NetworkXUnbounded("Negative cost cycle detected.")
+                            # Negative cycle found!
+                            if return_negative_cycle_info:
+                                return v, pred
+                            raise nx.NetworkXUnbounded("Negative cycle detected.")
```

Comment:
Now that the negative cycle search is toggled by a keyword argument, would it be easier to just run the function in that branch instead? In that case, maybe you could avoid exposing the `find_negative_cycles` function entirely - though maybe it's useful in other contexts?

This probably feels like the exact opposite of my original comment, but I think having it behind a new kwarg makes a big difference! The default behavior is the same as before (raise as soon as a negative cycle is detected), but now gives users the option to toggle a more detailed exception message if they wish. What do you think?

Reply:
If I understand correctly, you are suggesting we make a new boolean kwarg like:  `report_negative_cycles`
When False, (default) _bellman_ford raises an exception as soon as the cycle is detected. When True,
the extra work is done to find the cycle.

Does it make better sense to keep a separate (private) function to find the cycle, or define the helper function inside the `_bellman_ford` function itself.  I'm having trouble thinking of a use case for this function that doesn't require calling `_bellman_ford` first and collecting the `v, pred` variables from that function.

Reply:
> If I understand correctly, you are suggesting we make a new boolean kwarg like: report_negative_cycles
When False, (default) _bellman_ford raises an exception as soon as the cycle is detected. When True,
the extra work is done to find the cycle.

Yes this is what I was thinking - the main reason is that I found the `raise`/`return` behavior strange. Basically, if you hit the `if v in recent_update[u]:` branch, you've found a negative cycle, but depending on the value of the kwarg, you either get an exception *or* you completely change what `_bellman_ford` returns. This felt to me like something that might catch people out if they weren't reading the documentation very carefully. However; I don't see a way around it if you want to have `find_negative_cycle`, which requires that you somehow extract the `v, pred` info from `_bellman_ford`, as you point out.

Given all this, I think the current organization makes sense and `find_negative_cycle` seems useful to me (as evidenced by the motivating use-case), so this LGTM :+1: . The only remaining nit would be to update the `Returns` section of `_bellman_ford` to reflect the fact that it will return something different if a negative cycle is detected and `return_negative_cycle_info` is True.

Reply:
OK.... I've implemented the implementation I was thinking of. Basically, there is an inner private function that does bellman_ford relaxation. It updates the input parameter dicts `pred` and `dist` and returns the node where the relaxations stopped (usually None meaning it completed the process -- stopping early means a negative cycle was found).  The outer level private function calls the inner one, checks if it completed by examining the return value and then generates the `paths` dict if requested to do so.  This allows the new `find_negative_cycle` function to call the inner private function, look at the return value and the updated `pred` dict to find out whether there are any negative cycles and generate the negative cycle from the pred dict and the node at which the relaxations stopped.

This is ready for review.